hspec-core/src/Test/Hspec/Core/Extension.hs

```diff
@@ -0,0 +1,71 @@
+module Test.Hspec.Core.Extension (
+-- * Register command-line options
+  Option
+, OptionSetter
+, registerOption
+, flag
+, option
+, argument
+
+-- * Register tree transformations
+, runIO
+, modifyConfig
+, registerTransformation
+
+-- * Types
+-- |
+-- Operations on these types are provided by the following modules
+-- respectively:
+--
+-- * "Test.Hspec.Core.Extension.Config"
+-- * "Test.Hspec.Core.Extension.Item"
+-- * "Test.Hspec.Core.Extension.Spec"
+-- * "Test.Hspec.Core.Extension.Tree"
+, Config
+, Item
+, SpecM
+, SpecWith
+, SpecTree
+) where
+
+import           Prelude ()
+import           Test.Hspec.Core.Compat
+
+import qualified Data.CallStack as CallStack
+
+import qualified GetOpt.Declarative as Declarative
+import           Test.Hspec.Core.Spec (SpecM, SpecWith, runIO)
+import qualified Test.Hspec.Core.Spec as Core
+import qualified Test.Hspec.Core.Config.Definition as Core
+import           Test.Hspec.Core.Extension.Config (Config)
+import qualified Test.Hspec.Core.Extension.Config as Config
+import qualified Test.Hspec.Core.Extension.Config.Type as Config
+import           Test.Hspec.Core.Extension.Item (Item)
+import           Test.Hspec.Core.Extension.Tree (SpecTree)
+
+newtype Option = Option { unOption :: Declarative.Option Config }
+newtype OptionSetter = OptionSetter { unOptionSetter :: Declarative.OptionSetter Config }
+
+flag :: String -> (Bool -> Config -> Config) -> String -> Option
+flag name setter = Option . Core.flag name setter
+
+option :: String -> OptionSetter -> String -> Option
```

Comment:
https://github.com/hspec/hspec/pull/868/files#diff-89c0bd2feff5876f008665eb349641509884e5274c5eac99342be545d21d4bc3

Is this along the lines of what you had in mind?

Reply:
I'm not 100% solved on it, as it's not as clear anymore, which string is the option name, and which string is the name for the parameter.  But let's give this  a try, provisionally.

Reply:
You are right about the which string is which. Perhaps aliases would be helpful.
```haskell
argument :: CliName -> CliArgName -> CliHelp -> (Bool -> Config -> Config) -> Option
type CliName = String
type CliArgName = String
type CliHelp = String
```

I rearranged the arguments because I like having functions at the end which allows multiline lambdas without parenthaseses and hanging arguments. This should be applied to all the option functions if you choose to use it.